### DISCUSSION TITLE
Laravel 8.x custom log channel implementation with monolog

### DISCUSSION THREAD

**Discussion by karthi5053** [laravel/framework#34613] [General]
I trying to create a custom log handler in laravel using monolog with daily rotation. Please find the below configurations

in config/logging.php
'lbs' => [
            'driver' => 'daily',
            'via' => [App\Services\CustomLog::class],
            'level' => 'debug',
        ],

**in CustomLog.php Model**

`<?php
namespace App\Services;

use Monolog\Handler\RotatingFileHandler;
use Monolog\Handler\StreamHandler;
use Monolog\Logger;

class CustomLog
{
    public function __invoke($logger)
    {
        $suffix = date('Ymd') . '_' . php_sapi_name();
        $info_stream = new StreamHandler(storage_path("logs/lbs_info_".$suffix.".log"), Logger::INFO, false);
        $critical_stream = new StreamHandler(storage_path("logs/lbs_critical_".$suffix.".log"), Logger::CRITICAL, false);

        $logger->pushHandler($info_stream);
        $logger->pushHandler($critical_stream);
    }
}`
The above setup was working fine in laravel 5.8 version, but when i upgraded to laravel 8.3 this seems to be not working.

I am getting the below error

[2020-10-01 07:07:52] laravel.EMERGENCY: Unable to create configured logger. Using emergency logger. {"exception":"[object] (TypeError(code: 0): Argument 1 passed to Monolog\\Handler\\RotatingFileHandler::__construct() must be of the type string, null given, called in /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php on line 280 at /app/laravelbs2020/laravel83/laravel/vendor/monolog/monolog/src/Monolog/Handler/RotatingFileHandler.php:48)
[stacktrace]
#0 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(280): Monolog\\Handler\\RotatingFileHandler->__construct(NULL, 7, 100, true, NULL, false)
#1 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(202): Illuminate\\Log\\LogManager->createDailyDriver(Array)
#2 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(118): Illuminate\\Log\\LogManager->resolve('lbs')
#3 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(98): Illuminate\\Log\\LogManager->get('lbs')
#4 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(87): Illuminate\\Log\\LogManager->driver('lbs')
#5 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(241): Illuminate\\Log\\LogManager->channel('lbs')
#6 [internal function]: Illuminate\\Log\\LogManager->Illuminate\\Log\\{closure}('lbs', 0)
#7 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Collections/Collection.php(630): array_map(Object(Closure), Array, Array)
#8 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php(339): Illuminate\\Support\\Collection->map(Object(Closure))
#9 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(242): Illuminate\\Support\\Collection->flatMap(Object(Closure))
#10 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(202): Illuminate\\Log\\LogManager->createStackDriver(Array)
#11 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(118): Illuminate\\Log\\LogManager->resolve('stack')
#12 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(98): Illuminate\\Log\\LogManager->get('stack')
#13 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(533): Illuminate\\Log\\LogManager->driver()
#14 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(261): Illuminate\\Log\\LogManager->critical('This is critica...')
#15 /app/laravelbs2020/laravel83/laravel/app/Http/Controllers/TryoutController.php(16): Illuminate\\Support\\Facades\\Facade::__callStatic('critical', Array)
#16 [internal function]: App\\Http\\Controllers\\TryoutController->getCheckIt(Object(Illuminate\\Http\\Request))
#17 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)
#18 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\\Routing\\Controller->callAction('getCheckIt', Array)
#19 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Route.php(255): Illuminate\\Routing\\ControllerDispatcher->dispatch(Object(Illuminate\\Routing\\Route), Object(App\\Http\\Controllers\\TryoutController), 'getCheckIt')
#20 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Route.php(197): Illuminate\\Routing\\Route->runController()
#21 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php(691): Illuminate\\Routing\\Route->run()
#22 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))
#23 /app/laravelbs2020/laravel83/laravel/app/Http/Middleware/ForceSSL.php(26): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#24 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): App\\Http\\Middleware\\ForceSSL->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#25 /app/laravelbs2020/laravel83/laravel/app/Http/Middleware/AppBootstrap.php(35): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#26 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): App\\Http\\Middleware\\AppBootstrap->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#27 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(41): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#28 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#29 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#30 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#31 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#32 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(62): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest(Object(Illuminate\\Http\\Request), Object(Illuminate\\Session\\Store), Object(Closure))
#33 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Session\\Middleware\\StartSession->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#34 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#35 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#36 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#37 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#38 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#39 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php(693): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#40 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php(668): Illuminate\\Routing\\Router->runRouteWithinStack(Object(Illuminate\\Routing\\Route), Object(Illuminate\\Http\\Request))
#41 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php(634): Illuminate\\Routing\\Router->runRoute(Object(Illuminate\\Http\\Request), Object(Illuminate\\Routing\\Route))
#42 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Routing/Router.php(623): Illuminate\\Routing\\Router->dispatchToRoute(Object(Illuminate\\Http\\Request))
#43 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(166): Illuminate\\Routing\\Router->dispatch(Object(Illuminate\\Http\\Request))
#44 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}(Object(Illuminate\\Http\\Request))
#45 /app/laravelbs2020/laravel83/laravel/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php(60): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#46 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Barryvdh\\Debugbar\\Middleware\\InjectDebugbar->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#47 /app/laravelbs2020/laravel83/laravel/vendor/fruitcake/laravel-cors/src/HandleCors.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#48 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#49 /app/laravelbs2020/laravel83/laravel/app/Http/Middleware/AuditRequest.php(17): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#50 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): App\\Http\\Middleware\\AuditRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#51 /app/laravelbs2020/laravel83/laravel/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#52 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fideloper\\Proxy\\TrustProxies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#53 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#54 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#55 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#56 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#57 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#58 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#59 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(87): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#60 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#61 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#62 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(141): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#63 /app/laravelbs2020/laravel83/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#64 /app/laravelbs2020/laravel83/laravel/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#65 {main}

Let me know what could be the issue.

**Comment by mfn** [comment]
> Argument 1 passed to Monolog\Handler\RotatingFileHandler::__construct() must be of the type string, null given

This is fed from https://github.com/laravel/framework/blob/d2da49c113f995bce6ed64ac765ca3ab3fa6236c/src/Illuminate/Log/LogManager.php#L279 and as you can see their, your `$config` requires `path`, which you don't have in your config:

![image](https://user-images.githubusercontent.com/87493/94990015-313a0600-0579-11eb-861b-b494ef0754d9.png)

**Reply by karthi5053** [reply]
Hi Markus,

That helps. Thanks a lot..
